lib/Infer/AliveDriver.cpp

```diff
@@ -0,0 +1,406 @@
+    t.src = std::move(LHSF);
+    t.tgt = std::move(RHSF);
+    tools::TransformVerify tv(t, /*check_each_var=*/false);
+    int64_t result = 0;
+    if (auto errs = SynthesizeConstantUsingSolver(t, ConstName, result)) {
+      return std::nullopt; // TODO: Encode errs into ErrorCode
+    } else {
+      return {result};
+    }
+  } else {
+    assert(false && "Multiple constants unimplemented.");
+    return std::nullopt;
+  }
+}
+
+
+bool souper::AliveDriver::verify (Inst *RHS) {
+  if (!LHSFExists) return false;
+  RExprCache.clear();
+  IR::Function RHSF;
+  if (!translateRoot(RHS, nullptr, RHSF, RExprCache)) {
+    llvm::errs() << "Failed to translate RHS.\n";
+    // TODO: Eventually turn this into an assertion
+    return false;
+  }
+
+  tools::Transform t;
+  ReturnLHSRAII foo{t, LHSF};
+  t.src = std::move(LHSF);
+  t.tgt = std::move(RHSF);
+  tools::TransformVerify tv(t, /*check_each_var=*/false);
+
+  if (auto errs = tv.verify()) {
+    return false; // TODO: Encode errs into ErrorCode
+  } else {
+    llvm::errs() << "RHS proved valid.\n";
+    return true;
+  }
+}
+
+bool souper::AliveDriver::translateRoot(souper::Inst *I, Inst *PC,
+    IR::Function& F, Cache &ExprCache) {
+  if (!translateAndCache(I, F, ExprCache)) {
+    return false;
+  }
+  if (PC && !translateAndCache(PC, F, ExprCache)) {
+    return false;
+  }
+  FunctionBuilder Builder(F);
+  if (PC) {
+    Builder.Assume(ExprCache[PC]);
+  }
+  Builder.Ret(getType(I->Width), ExprCache[I]);
+  F.setType(getType(I->Width));
+  return true;
+}
+
+
+bool souper::AliveDriver::translateAndCache(souper::Inst *I,
+    IR::Function& F, Cache &ExprCache) {
+  if (ExprCache.find(I) != ExprCache.end()) {
+    return true; // Already translated
+  }
+
+  FunctionBuilder Builder(F);
+  for (auto &&Op : I->Ops) {
+    if(!translateAndCache(Op, F, ExprCache)) {
+      return false;
+    }
+  }
+
+  std::string Name = "";
+
+  if (NamesCache.find(I) != NamesCache.end()) {
+    Name = NamesCache[I];
+  } else if (I->Name != "") {
+    if (StartsWith("reservedconst_", I->Name)) {
+      // No way to avoid string matching without
+      // changes in Inst and ExhaustiveSynthesis
+      Name = "%" + I->Name;
+    } else {
+      Name = "%var_" + I->Name;
+    }
+  } else {
+    Name = "%" + std::to_string(InstNumbers++);
+    // FIXME: Somewhere the non-input variable names are discarded,
+    // forcing AliveDriver to name variables on its own.
+  }
+
+  auto &t = getType(I->Width);
+
+  switch (I->K) {
+    case souper::Inst::Var: {
+      ExprCache[I] = Builder.Var(t, Name);
+      return true;
+    }
+    case souper::Inst::Const: {
+      ExprCache[I] = Builder.Val(t, I->Val.getLimitedValue());
+      return true;
+    }
+
+    case souper::Inst::Select: {
+      ExprCache[I] = Builder.Select(t, Name,
+        ExprCache[I->Ops[0]],
+        ExprCache[I->Ops[1]],
+        ExprCache[I->Ops[2]]);
+    }
+
+    #define BINOP(SOUPER, ALIVE) case souper::Inst::SOUPER: {\
```

Comment:
would be best to align the backslashes, I forgot if you're an emacs user but emacs has a quick command for this (so no doubt vim also does)